src/kOS.Safe/Encapsulation/Structure.cs

```diff
@@ -337,20 +337,10 @@ public static Structure FromPrimitiveWithAssert(object value)
 
         public static object ToPrimitive(object value)
         {
-            var scalarValue = value as ScalarValue;
-            if (scalarValue != null)
+            var primitive = value as PrimitiveStructure;
+            if (primitive != null)
             {
-                return scalarValue.Value;
-            }
-            var booleanValue = value as BooleanValue;
-            if (booleanValue != null)
-            {
-                return booleanValue.Value;
-            }
-            var stringValue = value as StringValue;
-            if (stringValue != null)
-            {
-                return stringValue.ToString();
+                return primitive.ToPrimitive();
```

Comment:
OK, I may mess with this later to try and make it so inheriting from `PrimitiveStructure` can some how opt in to the `FromPrimitive` option.  This fixes half of the issue, because it unifies type serialization but doesn't unify de-serialization (unless I'm missing something, which I easily could, I really need to spend more time looking at serialization).